fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexScrubbing.java

```diff
@@ -0,0 +1,222 @@
+        tools.presetCommonParams(index, scrubbingPolicy.allowRepair(), common.getIndexContext().isSynthetic, common.getAllRecordTypes());
+
+        return rangeSet.firstMissingRangeAsync().thenCompose(range -> {
+            if (range == null) {
+                // Here: no more missing ranges - all done
+                // To avoid stale metadata, we'll keep the scrubbed-ranges indicator empty until the next scrub call.
+                rangeSet.clear();
+                return AsyncUtil.READY_FALSE;
+            }
+            final Tuple rangeStart = RangeSet.isFirstKey(range.begin) ? null : Tuple.fromBytes(range.begin);
+            final Tuple rangeEnd = RangeSet.isFinalKey(range.end) ? null : Tuple.fromBytes(range.end);
+            final TupleRange tupleRange = TupleRange.between(rangeStart, rangeEnd);
+
+            final RecordCursor<T> cursor = tools.getIterator(tupleRange, store, getLimit() + 1); // always respect limit in this path; +1 allows a continuation item in forward scan
+            final AtomicBoolean hasMore = new AtomicBoolean(true);
+            final AtomicReference<RecordCursorResult<T>> lastResult = new AtomicReference<>(RecordCursorResult.exhausted());
+            final long scanLimit = scrubbingPolicy.getEntriesScanLimit();
+            List<IndexScrubbingTools.Issue> issueList = new LinkedList<>();
+
+            return iterateRangeOnly(store, cursor, (recordStore, result) -> handleOneItem(recordStore, result, tools, issueList),
+                    lastResult, hasMore, recordsScanned, isIdempotent)
+                    .thenApply(vignore -> hasMore.get() ? tools.getContinuation(lastResult.get()) :
+                                          rangeEnd)
+                    .thenCompose(cont -> rangeSet.insertRangeAsync(packOrNull(rangeStart), packOrNull(cont), true)
+                            .thenApply(ignore -> notAllRangesExhausted(cont, rangeEnd)))
+                    .thenApply(ret -> {
+                        // Here: after a successful transaction, report the issues and check/update the scan counter.
+                        reportIssues(issueList);
+                        if (scanLimit > 0) {
+                            scanCounter += recordsScanned.get();
+                            if (scanLimit <= scanCounter) {
+                                return false;
+                            }
+                        }
+                        return ret;
+                    });
+        });
+    }
+
+
+    private <T> CompletableFuture<FDBStoredRecord<Message>> handleOneItem(FDBRecordStore store, final RecordCursorResult<T> result, final IndexScrubbingTools<T> tools, List<IndexScrubbingTools.Issue> issueList) {
+        return tools.handleOneItem(store,
+                        store.getIndexMaintainer(common.getIndex()).state.transaction,
+                        result)
+                .thenApply(issue -> {
```

Comment:
It will probably be easy to make `handleOneItem` return a nullable list of issues. But as you suggest - it's probably best to wait until this feature is needed.